exercises/src/main/scala/fpinscala/errorhandling/Option.scala

```diff
@@ -1,18 +1,27 @@
 package fpinscala.errorhandling
 
-
 import scala.{Option => _, Some => _, Either => _, _} // hide std library `Option`, `Some` and `Either`, since we are writing our own in this chapter
 
 sealed trait Option[+A] {
-  def map[B](f: A => B): Option[B] = sys.error("todo")
+  def map[B](f: A => B): Option[B] =
+    this match {
+      case Some(a) => Some(f(a))
+      case None => None
+    }
 
-  def getOrElse[B>:A](default: => B): B = sys.error("todo")
+  def getOrElse[B>:A](default: => B): B =
+    this match {
+      case Some(a) => a
+      case None => default
+    }
 
-  def flatMap[B](f: A => Option[B]): Option[B] = sys.error("todo")
+  def flatMap[B](f: A => Option[B]): Option[B] = map(f).getOrElse(None)
 
-  def orElse[B>:A](ob: => Option[B]): Option[B] = sys.error("todo")
+  def orElse[B>:A](ob: => Option[B]): Option[B] =
+    map(Some(_)).getOrElse(ob)
```

Comment:
@gpoirier I need to return something of type `Option[B]`, which has to always be actually a `Some`. Leaving out the `map` doesn't typecheck because `this.getOrElse(ob)` is a type conflict: `A` from the 'get' branch and `Option[B]` from the 'or else' branch.